// @aws-sdk/client-ivs-realtime/dist-cjs/IVSRealTime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IVSRealTime = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const CreateEncoderConfigurationCommand_1 = require("./commands/CreateEncoderConfigurationCommand");
const CreateParticipantTokenCommand_1 = require("./commands/CreateParticipantTokenCommand");
const CreateStageCommand_1 = require("./commands/CreateStageCommand");
const CreateStorageConfigurationCommand_1 = require("./commands/CreateStorageConfigurationCommand");
const DeleteEncoderConfigurationCommand_1 = require("./commands/DeleteEncoderConfigurationCommand");
const DeleteStageCommand_1 = require("./commands/DeleteStageCommand");
const DeleteStorageConfigurationCommand_1 = require("./commands/DeleteStorageConfigurationCommand");
const DisconnectParticipantCommand_1 = require("./commands/DisconnectParticipantCommand");
const GetCompositionCommand_1 = require("./commands/GetCompositionCommand");
const GetEncoderConfigurationCommand_1 = require("./commands/GetEncoderConfigurationCommand");
const GetParticipantCommand_1 = require("./commands/GetParticipantCommand");
const GetStageCommand_1 = require("./commands/GetStageCommand");
const GetStageSessionCommand_1 = require("./commands/GetStageSessionCommand");
const GetStorageConfigurationCommand_1 = require("./commands/GetStorageConfigurationCommand");
const ListCompositionsCommand_1 = require("./commands/ListCompositionsCommand");
const ListEncoderConfigurationsCommand_1 = require("./commands/ListEncoderConfigurationsCommand");
const ListParticipantEventsCommand_1 = require("./commands/ListParticipantEventsCommand");
const ListParticipantsCommand_1 = require("./commands/ListParticipantsCommand");
const ListStagesCommand_1 = require("./commands/ListStagesCommand");
const ListStageSessionsCommand_1 = require("./commands/ListStageSessionsCommand");
const ListStorageConfigurationsCommand_1 = require("./commands/ListStorageConfigurationsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const StartCompositionCommand_1 = require("./commands/StartCompositionCommand");
const StopCompositionCommand_1 = require("./commands/StopCompositionCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateStageCommand_1 = require("./commands/UpdateStageCommand");
const IVSRealTimeClient_1 = require("./IVSRealTimeClient");
const commands = {
    CreateEncoderConfigurationCommand: CreateEncoderConfigurationCommand_1.CreateEncoderConfigurationCommand,
    CreateParticipantTokenCommand: CreateParticipantTokenCommand_1.CreateParticipantTokenCommand,
    CreateStageCommand: CreateStageCommand_1.CreateStageCommand,
    CreateStorageConfigurationCommand: CreateStorageConfigurationCommand_1.CreateStorageConfigurationCommand,
    DeleteEncoderConfigurationCommand: DeleteEncoderConfigurationCommand_1.DeleteEncoderConfigurationCommand,
    DeleteStageCommand: DeleteStageCommand_1.DeleteStageCommand,
    DeleteStorageConfigurationCommand: DeleteStorageConfigurationCommand_1.DeleteStorageConfigurationCommand,
    DisconnectParticipantCommand: DisconnectParticipantCommand_1.DisconnectParticipantCommand,
    GetCompositionCommand: GetCompositionCommand_1.GetCompositionCommand,
    GetEncoderConfigurationCommand: GetEncoderConfigurationCommand_1.GetEncoderConfigurationCommand,
    GetParticipantCommand: GetParticipantCommand_1.GetParticipantCommand,
    GetStageCommand: GetStageCommand_1.GetStageCommand,
    GetStageSessionCommand: GetStageSessionCommand_1.GetStageSessionCommand,
    GetStorageConfigurationCommand: GetStorageConfigurationCommand_1.GetStorageConfigurationCommand,
    ListCompositionsCommand: ListCompositionsCommand_1.ListCompositionsCommand,
    ListEncoderConfigurationsCommand: ListEncoderConfigurationsCommand_1.ListEncoderConfigurationsCommand,
    ListParticipantEventsCommand: ListParticipantEventsCommand_1.ListParticipantEventsCommand,
    ListParticipantsCommand: ListParticipantsCommand_1.ListParticipantsCommand,
    ListStagesCommand: ListStagesCommand_1.ListStagesCommand,
    ListStageSessionsCommand: ListStageSessionsCommand_1.ListStageSessionsCommand,
    ListStorageConfigurationsCommand: ListStorageConfigurationsCommand_1.ListStorageConfigurationsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    StartCompositionCommand: StartCompositionCommand_1.StartCompositionCommand,
    StopCompositionCommand: StopCompositionCommand_1.StopCompositionCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateStageCommand: UpdateStageCommand_1.UpdateStageCommand,
};
class IVSRealTime extends IVSRealTimeClient_1.IVSRealTimeClient {
}
exports.IVSRealTime = IVSRealTime;
(0, smithy_client_1.createAggregatedClient)(commands, IVSRealTime);
